File: src/main/java/videostore/horror/Customer.java
package videostore.horror;

import java.util.*;

import static videostore.horror.PriceCode.NEW_RELEASE;

record Rental {
    Movie movie;
    int daysRented;
}

class Customer {
    private final String name;
    private Map<Movie, Integer> rentals = new LinkedHashMap<>(); // preserves order

    public Customer(String name) {
        this.name = name;
    }

    ;

    public void addRental(Movie m, int d) {
        rentals.put(m, d);
    }

    public String getName() {
        return name;
    }

    // test
    public String statement() {
        double totalPrice = 0;
        int frequentRenterPoints = 0;
        String result = "Rental Record for " + getName() + "\n";
        for (Movie movie : rentals.keySet()) {
            int daysRented = rentals.get(movie);
            totalPrice += calculatePrice(movie.priceCode(), daysRented);

            frequentRenterPoints += calculatePoints(movie, daysRented);

            // show figures line for this rental
            result += "\t" + movie.title() + "\t" + calculatePrice(movie.priceCode(), daysRented) + "\n";
        }
        // add footer lines
        result += "Amount owed is " + totalPrice + "\n";
        result += "You earned " + frequentRenterPoints + " frequent renter points";
        return result;
    }

    private static int calculatePoints(Movie movie, int daysRental) {
        int frequentRenterPoints = 1;
        if ((movie.priceCode() == NEW_RELEASE) && daysRental >= 2)
            frequentRenterPoints++;
        return frequentRenterPoints;
    }

    private static double calculatePrice(PriceCode priceCode, final int daysRented) {
        switch (priceCode) {
            case REGULAR -> {
                return 2 + ((daysRented > 2) ? (daysRented - 2) * 1.5 : 0);
            }
            case NEW_RELEASE -> {
                return daysRented * 3;
            }
            case CHILDREN -> {
                double price = 1.5;
                if (daysRented > 3)
                    price += (daysRented - 3) * 1.5;
                return price;
            }
            default -> throw new IllegalStateException("Unexpected value: " + priceCode);
        }
    }
}

File: cs/1_Original/Soat.CleanCode.VideoStore.Original.TypeScript/src/customer.ts
import {Movie} from './movie';
import {Rental} from './rental';

export class Customer {
    private _rentals: Rental[] = [];

    constructor(readonly name: string) {
    }

    addRental(rental: Rental) {
        this._rentals.push(rental);
    }

    statement(): string {
        let frequentRenterPoints = 0;
        let totalAmount = 0;
        let result = 'Rental Record for ' + this.name + '\n';
        for (let each of this._rentals) {
            let thisAmount = 0;

            // dtermines the amount for each line
            switch (each.movie.priceCode) {
                case Movie.REGULAR:
                    thisAmount += 2;
                    if (each.daysRented > 2) {
                        thisAmount = thisAmount + (each.daysRented - 2) * 1.5;
                    }
                    break;
                case Movie.NEW_RELEASE:
                    thisAmount = thisAmount + each.daysRented * 3;
                    break;
                case Movie.CHILDREN:
                    thisAmount += 1.5;
                    if (each.daysRented > 3) {
                        thisAmount = thisAmount + (each.daysRented - 3) * 1.5;
                    }
                    break;
            }

            frequentRenterPoints++;
            if (each.movie.priceCode == Movie.NEW_RELEASE && each.daysRented > 1) {
                frequentRenterPoints++;
            }

            result += '\t' + each.movie.title + '\t' + thisAmount.toFixed(1) + '\n';
            totalAmount += thisAmount;
        }

        result += 'You owed ' + totalAmount.toFixed(1) + '\n';
        result += 'You earned ' + frequentRenterPoints + ' frequent renter points \n';

        return result;
    }
}